### ISSUE TITLE
UpstreamPathTemplate hard-coded priority 0

### ISSUE THREAD

**Issue by davidni** [ThreeMammals/Ocelot#479]
## Expected Behavior / New Feature
`UpstreamTemplatePatternCreator.cs:33` creates an instance of `UpstreamPathTemplate` but sets priority to `0` instead of using `reRoute.Priority`. I haven't validated whether this affects behavior, but it seems like a potential bug.

## Specifications

  - Version: 8.0.0

**Comment by TomPallister** [comment]
@davidni its probably a mistake that doesnt affect the logic. Will check ASAP. Fixing a bunch of bugs atm 😄

**Comment by TomPallister** [comment]
This is intentional because I've hardcoded the use case /{url} to always be the lowest priority which i think is OK for most cases. However in the future I guess this could be controlled by users...not sure...

 I'm not super happy with how the config translates to the functionality at the moment...but Im not sure how it can be made better without massively over complicating it.

I should probably just let people set their own regex here in addition to the templating system Ocelot provides.

**Comment by davidni** [comment]
Isn't this problematic for routes with different UpstreamHosts? Or at least it would be once I've added support for wildcards in host names. Overriding the priority of a route based on its pathTemplate seems undesirable.

Side note: it seems in fact that there are still assumptions in place from before Ocelot supported custom UpstreamHost names. For example, the fluent config validator tries to enforce that there can be no 2 rules with the same path template, but I could have a rule for `/{everything}` on `a.com` and a separate rule for `/{everything}` on `b.com`. I have had to disable the validator temporarily to enable this scenario.

**Comment by TomPallister** [comment]
@davidni yep that makes sense :E

**Comment by TomPallister** [comment]
@davidni i checked the code I think it works ok if you have two /upstream path templates with different hosts.

```csharp
 var matchingReRoutes = reRoutes
                .Where(r => r.UpstreamPathTemplate == reRoute.UpstreamPathTemplate 
                            && (r.UpstreamHost != reRoute.UpstreamHost || reRoute.UpstreamHost == null))
                .ToList();
```

Then I added this test to make sure

```csharp
[Fact]
        public void test()
        {
            this.Given(x => x.GivenAConfiguration(new FileConfiguration
            {
                ReRoutes = new List<FileReRoute>
                {
                    new FileReRoute
                    {
                        DownstreamPathTemplate = "/api/products/",
                        UpstreamPathTemplate = "/asdf/",
                        DownstreamHostAndPorts = new List<FileHostAndPort>
                        {
                            new FileHostAndPort
                            {
                                Host = "bbc.co.uk"
                            }
                        },
                        UpstreamHost = "bbc.co.uk"
                    },
                    new FileReRoute
                    {
                        DownstreamPathTemplate = "/api/products/",
                        UpstreamPathTemplate = "/asdf/",
                        DownstreamHostAndPorts = new List<FileHostAndPort>
                        {
                            new FileHostAndPort
                            {
                                Host = "bbc.co.uk"
                            }
                        },
                        UpstreamHost = "itv.co.uk"
                    }
                }
            }))
                .When(x => x.WhenIValidateTheConfiguration())
                .Then(x => x.ThenTheResultIsValid())
                .BDDfy();
        }
```

I guess the problem you will have is if it is possible to match multiple hosts with your wildcard and you have that situation occur Ocelot will choose the first listed in the ReRoutes config unless you override with priority.

**Comment by davidni** [comment]
Here's what I was thinking of:

```json
"ReRoutes": [
  {
    "UpstreamHost": "*.abc.com",
    "UpstreamPathTemplate": "/{everything}",
    "DownstreamPathTemplate": "/foo/{everything}",
    "Priority": 9,
    /* ... */
  }, {
    "UpstreamHost": "a.abc.com",
    "UpstreamPathTemplate": "/{everything}",
    "DownstreamPathTemplate": "/bar/{everything}",
    "Priority": 10,
    /* ... */
  }
]```

If my understanding is correct, priority is ignored for the routes above (because UpstreamPathTemplate is a catch all). Which results in the first rule being applied first, which will effectively hide the other rule. Whereas the desired behavior was the opposite since the second rule has a higher priority.

**Comment by TomPallister** [comment]
@davidni yep ok that does make sense, it is a simple change. Do you want to change it or shall I?

**Comment by davidni** [comment]
Update: I think it will make sense for me to tackle this as part of #480 since it involves rewriting the rule ordering and matching code anyway

**Comment by AlyHKafoury** [comment]
@TomPallister @davidni Can I work on this or is this being worked on ?

**Comment by raman-m** [comment]
> @AlyHKafoury commented [on Jul 13, 2023](#issuecomment-1634629211)

Hi Aly!
Thanks for your interest in Ocelot!
Sure you can prepare some PR to review.
Both issues, [this]() one, #479 and #480, are still not accepted by organization members including Tom.
Sometime Tom gets an issue or PR an work in collaboration mode or personally. But he doesn't work actively on this his project since 2020.
So, I would say that you can wait for Tom's accepting or mine. To be accepted by someone who is going to write the code really, not virtually, and review, and merge.

From my side I can accept the issue if a PR will be prepared.
So, as a member of Ocelot team I try to follow the rule: **"If a community member has prepared a PR for an issue then this issue will be auto accepted having accepted label for this PR"**.

Mr. David Nissimoff, who was mentioned by you, he has prepared **0 (zero)** PRs last 5 years as for this issue and for his created issue #480. I would say he is not interested to contribute to this project. 
Who knows? Maybe he will answer to you...

About Tom...
Just ask Tom directedly!  😄

**Comment by raman-m** [comment]
> @davidni commented [on Jul 26, 2018](#issuecomment-408219904)

Hey David!
How things going last 5 years? 

Issue #480 is about wildcard hostnames of upstream, but this issue is about priorities of the routes.
My opinion: they are completely different features!
It has zero efficiency to discuss dependencies between priority and wildcard features until at least one of them will be implemented and merged as a PR.

Do you have an intention to contribute to this project?
Depending on your future answer I need to decide what do to with your assigning to this issue. Tom had a hope you implement the feature.

**Comment by raman-m** [comment]
@davidni commented [on Jul 26, 2018](#issuecomment-408208104):
> Please go ahead if you can. I'll focus on #480 (wildcards in hostnames) today.

You were working on #480 ... Why this issue is still open in 2023 ?
What's the PR number?

**Comment by AlyHKafoury** [comment]
@raman-m Can I continue on this task, I am still interested

**Comment by raman-m** [comment]
> @raman-m Can I continue on this task, I am still interested

@AlyHKafoury Sure, you can!
But focus on the bug with priority only, plz!
The wildcard feature should be ignored for now.
So, a bugfix for priority shouldn't have complex code. Please care about unit & acceptance tests!

**Comment by AlyHKafoury** [comment]
@raman-m I have suggestion that I believe implement can solve the problem without changing the unit tests a lot and will still allow priorities for for placeholder paths to work i.e.: "/{url}" is to use negative priority numbers internally for them, that we they won't collide with normal priorities and in the same time their priorities will still be taken into consideration

**Comment by AlyHKafoury** [comment]
That is to avoid breaking things like [Ocelot.AcceptanceTests/RoutingTests.cs#L151 ](https://github.com/ThreeMammals/Ocelot/blob/bb795876c493af958f1d58c852ea56e94263906d/test/Ocelot.AcceptanceTests/RoutingTests.cs#L151)

```

**Comment by raman-m** [comment]
@AlyHKafoury Your implementation should
- **not** break current features
- review current tests and changing them possibly without breaking features
- consider CatchAll route with lowest priority `0`
- consider non-CatchAll routes with custom priority level

This is my understanding.

**Comment by AlyHKafoury** [comment]
@raman-m I've tested multiple test cases manually, the bug only exists if you consider implementing #480 if we don't implement that then there is no bug to consider her the __CatchAll__ routes gets priority of `0` and other routes gets correct priority from config

**Comment by AlyHKafoury** [comment]
@raman-m I am interested in implementing #480 if you are ok with it also, waiting for your confirmation
